# SSC Data Science and Analytics Workshop 2022

### Intro to Databases in Industry: Data Cleaning, Querying, and Modeling at Scale
-----------------

# 1. Introduction

### 1.1 Workshop structure

The workshop will be split in two parts. 

- In the first part, Arman and I will discuss
    - what it is and why to use a Database Management Systems (DBMS);
    - the basics of SQL queries;
    - accessing a database from R;
    
    
    
- In the second part, Diego will:
   - discuss the concept of a data warehouse;
   - introduce the ETL (Extract, Transform, Load) and ELT (Extract, Load, Transform) processes;
   - present dimensional modelling as methodology to guide the design of data warehouses at scale.
   
The workshop will be focused on the user aspect of databases, not on design and implementation. 

### 1.2 Why use a Database management system (DBMS)?

To analyze data, we need to first collect and store the data. In fact, depending on the scale of the dataset, we need much more than that. Curating a large dataset is not just about storing it, we also need to:

- keep track of the relationship between the data entities;
- avoid redundancies; 
- manage security issues to protect private and sensitive information;
- have an error recovery system in place;
- be able to retrieve data in an efficient manner;
- have well controlled concurrent access;
- keep the data in a consistent state;
- enforce data integrity!

### 1.3 Database vs DBMS 

- A **database is a collection of data about different but related objects** (also called entities in database jargon), each entity with a common set of attributes. For example, a taxi company stores information about:
  - Drivers (name, licence number, licence expiration);
  - Cars (model, year);
  - Clients (name, address, phone, e-mail);


- Note that entities also interact among themselves (e.g., a driver drives a car).

- A **DBMS is a software** responsible to assist the maintenance and usage of databases (yes, plural!)
  - There are many different vendors of **relational** DBMS out there (e.g., PostgreSQL, MS Access, MS SQL Server, MySQL, etc…)
  <img src="img/lecture1/flavours_sql.png" width="700"/>

---

**Remember:**
    
database $\ne$ database management system

---

### 1.4 PostgreSQL

- In this workshop we will be using PostgreSQL as our RDBMS;


- PostgreSQL is an open-source multi-platform (Windows, Linux, macOS) RDBMS with a high level of compliance with ANSI SQL (the langugae that practically all RDBMS use to create, modify and query databases);

### 1.5 The client-server model

Similar to most other DBMSs, Postgres works based on a **client-sever** model. In this model

- The DBMS along with its databases and data are all stored on a host computer where the database server resides. This is typically a powerful machine with high processing power and large storage
- Client hosts are usually personal computers with GUIs that can connect to a database server to access the data.

In this model, the clients and the server are connected over a network. The heavy-lifting of processing, managing and storing large amounts of data is done by the server host, and clients only retrieve the data that they need.

<img src="img/lecture1/client-server.png" width="500"/>

> Although sometimes used interchangeably, there is a difference between a **client/server** and a **client/server host**. A host is a device, whereas a client/server is a piece of software. For example, you can simultaneously have multiple client programs connected to a remote database. Similarly, a remote host is a device (i.e. a computer) that might have several server programs running concurrently.

The idea of client-server models for databases has become the standard of computation and storage today, known as **cloud computing**:

- Today we rarely store movie or music files on our computers. This is why most of us have laptops with only 256/512 GB of space, because most of that takes up space is already provided as a cloud service (e.g. Netflix, Spotify, Youtube), or is stored on cloud storage spaces (e.g. One Drive, Dropbox, Google Drive).
- We rarely run production-stage computation-intensive jobs on our own computers. All such computations are done on cloud-computing services (e.g. Google Cloud Platform, Amazon Web Services, Microsoft Azure). I personally haven't run a single simulation code on my own computer, neither ever stored any raw data locally. I use my computer mainly as an interface to access the services that I want.

> Note that there are certain situations where one might want to **locally** benefit from the advantages of storing data in a database. A relational database engine that works only with local databases is SQLite. If you're curious to find out the use cases for **SQLite**, take a look [here](https://www.sqlite.org/whentouse.html).

Whenever we use Postgres (or any other client-server DBMS), the first step before anything else is to **connect** to the database server. This is why we will talk about _host address_, _port_, _username_, and _password_ when we try to use a database.

### 1.5 What is SQL?

Well, it's finally time to learn about SQL!

- SQL stands for Structured Query Language ([or... does it?](https://en.wikipedia.org/wiki/SQL#History)).

- It is a language that we use to talk to a relational DBMS.

- Originally developed by IBM in 1970s to manipulate and retrieve data stored in their DBMS, System R.

- Keep in mind that SQL is not set in stone and the SQL standard is updated from time to time (e.g., 1986, 1992, 1999, 2011, 2016);

- There’s no DBMS that fully complies with the standard;

## 2. SQL

Suppose that we have the following table in our database, and 

> we want to retrieve the names and GPAs of students older than 25.

|  sid  | name      | login      | age | gpa |
|-------|-----------|------------|-----|-----|
| 23792 | Arman     | arman@mds  | 28  | 2.5 |
| 82347 | Varada    | varada@mds | 29  | 2.9 |
| 11238 | Tiffany   | tiff@mds   | 23  | 2.8 |
| 87263 | Mike      | mike@mds   | 19  | 3.8 |
| 13298 | Joel      | joel@mds   | 25  | 3.2 |
| 91287 | Florencia | flor@mds   | 20  | 3.3 |

We can write this as the following SQL query:

```sql
SELECT
    name, age, gpa
FROM
    Students
WHERE
    age > 25;
```

Running the above query should return this relation:

| name   | age | gpa |
|--------|-----|-----|
| Arman  | 28  | 2.5 |
| Varada | 29  | 2.9 |

Let's dissect the different parts of our SQL query here.

### 2.1 The `SELECT` statement

```sql
-- A basic SELECT statement
-- ===========================
SELECT
    name, age, gpa  -- column names
FROM
    Students        -- table name
WHERE
    age > 25;       -- condition
```

- A SQL statement usually starts with a verb that describes what the statement is doing, e.g., `SELECT`, `UPDATE`, `INSERT`, `CREATE`, …

- SQL is NOT case sensitive: “SELECT”, “select”, and “SeLeCt” are all the same;

- A SQL command can use multiple lines;

- Don't forget that every SQL statement needs to be terminated with a `;`.

- SQL keywords are traditionally written in upper case letters, but that is not a requirement.

### 2.1 How to run SQL in Postgres?

Well, we have a variety of options to run our SQL statements in PostgreSQL:

- pgAdmin is the official web-based GUI for interacting with PostgreSQL databases
- `psql` is PostgreSQL's interactive command-line interface
- `%sql` and `%%sql` magic commands in Jupyter notebooks, which are provided by the `ipython-sql` package
- `DBI` and `RPostgres` packages in R;

---------------------------
### 2.1.1 `psql`

This is PostgreSQL's command-line tool that allows us to interactively run SQL statements as well as "meta" commands. I introduce a couple of useful `psql` meta commands here, but you can find all the other ones in Postgres documentations [here](https://www.postgresql.org/docs/current/app-psql.html) or a shorter version in this [cheatsheet](http://www.postgresonline.com/downloads/special_feature/postgresql83_psql_cheatsheet.pdf).

| Command | Usage                                         |
|---------|-----------------------------------------------|
| `\l`    | list all databases                            |
| `\c`    | connect to a database                         |
| `\cd`   | change directory                              |
| `\!`    | execute shell commands                        |
| `\i`    | execute commands from file                    |
| `\d`    | list tables and views                         |
| `\d+`   | list tables and views with additional info    |
| `\dt`   | list tables                                   |
| `\dt+`  | list tables with additional info              |
| `\h`    | view help on SQL commands                     |
| `\?`    | view help on psql meta commands               |
| `\q`    | quit interactive shell                        |

> Note that you don't need to terminate meta commands with `;`.

### 2.1.2 `ipython-sql` (`%sql` and `%%sql`)

`ipython-sql` is a package that enables us to run SQL statements right from a Jupyter notebook. In order to use it, we should load it first:

In [49]:
%load_ext sql
%config SqlMagic.displaylimit = 30

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


Now we need the host address of where the database is stored, along with a username and a password.

It is always a bad idea to store login information directly in a notebook or code file because of security reasons. For example, you don't want to commit your sensitive login information to a Git repo.

In order to avoid that, we store that kind of information in a separate file, like `credentials.json` here, and read the username and password into our IPython session:

In [51]:
import json

with open('credentials.json') as f:
    credentials = json.load(f)
    
username = credentials['user']
password = credentials['password']
host = credentials['host']
port = credentials['port']

And also make sure to add your file name (e.g. `credentials.json`) to your `.gitignore` file, so you don't accidentally commit it.

Now we can establish the connection to the `world` database using the following code:

In [52]:
%sql postgresql://{username}:{password}@{host}:{port}/world

Note that we have used the `%sql` line magic to interpret the line in front of it as a magic command. 

We can also use `%%sql` cell magic to apply the magic to an entire notebook cell.

A limited number of `psql` meta commands (e.g. `\l`, `d`) can also be executed here. This is made possible through the `pgspecial` package. For example, let's list all databases that exist on our PostgreSQL server:

In [4]:
#%sql \l

Or list the relations (i.e. tables) in the current database:

In [5]:
#%sql \d

---------------------------------------
#### Wait, isn't this a workshop? Let's get to work then!!
To learn you need to practice! 

**Exercise 1:**

Retrieve the `name` and `population` columns from the `country` table:

In [53]:
%%sql 

SELECT name, population 
    FROM country;

   postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/imdb
 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
239 rows affected.


[('Afghanistan', 22720000),
 ('Netherlands', 15864000),
 ('Netherlands Antilles', 217000),
 ('Albania', 3401200),
 ('Algeria', 31471000),
 ('American Samoa', 68000),
 ('Andorra', 78000),
 ('Angola', 12878000),
 ('Anguilla', 8000),
 ('Antigua and Barbuda', 68000),
 ('United Arab Emirates', 2441000),
 ('Argentina', 37032000),
 ('Armenia', 3520000),
 ('Aruba', 103000),
 ('Australia', 18886000),
 ('Azerbaijan', 7734000),
 ('Bahamas', 307000),
 ('Bahrain', 617000),
 ('Bangladesh', 129155000),
 ('Barbados', 270000),
 ('Belgium', 10239000),
 ('Belize', 241000),
 ('Benin', 6097000),
 ('Bermuda', 65000),
 ('Bhutan', 2124000),
 ('Bolivia', 8329000),
 ('Bosnia and Herzegovina', 3972000),
 ('Botswana', 1622000),
 ('Brazil', 170115000),
 ('United Kingdom', 59623400),
 ('Virgin Islands, British', 21000),
 ('Brunei', 328000),
 ('Bulgaria', 8190900),
 ('Burkina Faso', 11937000),
 ('Burundi', 6695000),
 ('Cayman Islands', 38000),
 ('Chile', 15211000),
 ('Cook Islands', 20000),
 ('Costa Rica', 4023000),
 ('Djibouti', 638000),
 ('Dominica', 71000),
 ('Dominican Republic', 8495000),
 ('Ecuador', 12646000),
 ('Egypt', 68470000),
 ('El Salvador', 6276000),
 ('Eritrea', 3850000),
 ('Spain', 39441700),
 ('South Africa', 40377000),
 ('Ethiopia', 62565000),
 ('Falkland Islands', 2000),
 ('Fiji Islands', 817000),
 ('Philippines', 75967000),
 ('Faroe Islands', 43000),
 ('Gabon', 1226000),
 ('Gambia', 1305000),
 ('Georgia', 4968000),
 ('Ghana', 20212000),
 ('Gibraltar', 25000),
 ('Grenada', 94000),
 ('Greenland', 56000),
 ('Guadeloupe', 456000),
 ('Guam', 168000),
 ('Guatemala', 11385000),
 ('Guinea', 7430000),
 ('Guinea-Bissau', 1213000),
 ('Guyana', 861000),
 ('Haiti', 8222000),
 ('Honduras', 6485000),
 ('Hong Kong', 6782000),
 ('Svalbard and Jan Mayen', 3200),
 ('Indonesia', 212107000),
 ('India', 1013662000),
 ('Iraq', 23115000),
 ('Iran', 67702000),
 ('Ireland', 3775100),
 ('Iceland', 279000),
 ('Israel', 6217000),
 ('Italy', 57680000),
 ('East Timor', 885000),
 ('Austria', 8091800),
 ('Jamaica', 2583000),
 ('Japan', 126714000),
 ('Yemen', 18112000),
 ('Jordan', 5083000),
 ('Christmas Island', 2500),
 ('Yugoslavia', 10640000),
 ('Cambodia', 11168000),
 ('Cameroon', 15085000),
 ('Canada', 31147000),
 ('Cape Verde', 428000),
 ('Kazakstan', 16223000),
 ('Kenya', 30080000),
 ('Central African Republic', 3615000),
 ('China', 1277558000),
 ('Kyrgyzstan', 4699000),
 ('Kiribati', 83000),
 ('Colombia', 42321000),
 ('Comoros', 578000),
 ('Congo', 2943000),
 ('Congo, The Democratic Republic of the', 51654000),
 ('Cocos (Keeling) Islands', 600),
 ('North Korea', 24039000),
 ('South Korea', 46844000),
 ('Greece', 10545700),
 ('Croatia', 4473000),
 ('Cuba', 11201000),
 ('Kuwait', 1972000),
 ('Cyprus', 754700),
 ('Laos', 5433000),
 ('Latvia', 2424200),
 ('Lesotho', 2153000),
 ('Lebanon', 3282000),
 ('Liberia', 3154000),
 ('Libyan Arab Jamahiriya', 5605000),
 ('Liechtenstein', 32300),
 ('Lithuania', 3698500),
 ('Luxembourg', 435700),
 ('Western Sahara', 293000),
 ('Macao', 473000),
 ('Madagascar', 15942000),
 ('Macedonia', 2024000),
 ('Malawi', 10925000),
 ('Maldives', 286000),
 ('Malaysia', 22244000),
 ('Mali', 11234000),
 ('Malta', 380200),
 ('Morocco', 28351000),
 ('Marshall Islands', 64000),
 ('Martinique', 395000),
 ('Mauritania', 2670000),
 ('Mauritius', 1158000),
 ('Mayotte', 149000),
 ('Mexico', 98881000),
 ('Micronesia, Federated States of', 119000),
 ('Moldova', 4380000),
 ('Monaco', 34000),
 ('Mongolia', 2662000),
 ('Montserrat', 11000),
 ('Mozambique', 19680000),
 ('Myanmar', 45611000),
 ('Namibia', 1726000),
 ('Nauru', 12000),
 ('Nepal', 23930000),
 ('Nicaragua', 5074000),
 ('Niger', 10730000),
 ('Nigeria', 111506000),
 ('Niue', 2000),
 ('Norfolk Island', 2000),
 ('Norway', 4478500),
 ('Côte d\x92Ivoire', 14786000),
 ('Oman', 2542000),
 ('Pakistan', 156483000),
 ('Palau', 19000),
 ('Panama', 2856000),
 ('Papua New Guinea', 4807000),
 ('Paraguay', 5496000),
 ('Peru', 25662000),
 ('Pitcairn', 50),
 ('Northern Mariana Islands', 78000),
 ('Portugal', 999760

**Exercise 2: `SELECT *`?**

What happens if you replace your column list with * in the SELECT statement?

In [7]:
%%sql 

SELECT *
    FROM country;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
239 rows affected.


[('AFG', 'Afghanistan', 'Asia', 'Southern and Central Asia', 652090.0, 1919, 22720000, 45.9, Decimal('5976.00'), None, 'Afganistan/Afqanestan', 'Islamic Emirate', 'Mohammad Omar', 1, 'AF'),
 ('NLD', 'Netherlands', 'Europe', 'Western Europe', 41526.0, 1581, 15864000, 78.3, Decimal('371362.00'), Decimal('360478.00'), 'Nederland', 'Constitutional Monarchy', 'Beatrix', 5, 'NL'),
 ('ANT', 'Netherlands Antilles', 'North America', 'Caribbean', 800.0, None, 217000, 74.7, Decimal('1941.00'), None, 'Nederlandse Antillen', 'Nonmetropolitan Territory of The Netherlands', 'Beatrix', 33, 'AN'),
 ('ALB', 'Albania', 'Europe', 'Southern Europe', 28748.0, 1912, 3401200, 71.6, Decimal('3205.00'), Decimal('2500.00'), 'Shqipëria', 'Republic', 'Rexhep Mejdani', 34, 'AL'),
 ('DZA', 'Algeria', 'Africa', 'Northern Africa', 2381741.0, 1962, 31471000, 69.7, Decimal('49982.00'), Decimal('46966.00'), 'Al-Jaza\x92ir/Algérie', 'Republic', 'Abdelaziz Bouteflika', 35, 'DZ'),
 ('ASM', 'American Samoa', 'Oceania', 'Polynesia', 199.0, None, 68000, 75.1, Decimal('334.00'), None, 'Amerika Samoa', 'US Territory', 'George W. Bush', 54, 'AS'),
 ('AND', 'Andorra', 'Europe', 'Southern Europe', 468.0, 1278, 78000, 83.5, Decimal('1630.00'), None, 'Andorra', 'Parliamentary Coprincipality', '', 55, 'AD'),
 ('AGO', 'Angola', 'Africa', 'Central Africa', 1246700.0, 1975, 12878000, 38.3, Decimal('6648.00'), Decimal('7984.00'), 'Angola', 'Republic', 'José Eduardo dos Santos', 56, 'AO'),
 ('AIA', 'Anguilla', 'North America', 'Caribbean', 96.0, None, 8000, 76.1, Decimal('63.20'), None, 'Anguilla', 'Dependent Territory of the UK', 'Elisabeth II', 62, 'AI'),
 ('ATG', 'Antigua and Barbuda', 'North America', 'Caribbean', 442.0, 1981, 68000, 70.5, Decimal('612.00'), Decimal('584.00'), 'Antigua and Barbuda', 'Constitutional Monarchy', 'Elisabeth II', 63, 'AG'),
 ('ARE', 'United Arab Emirates', 'Asia', 'Middle East', 83600.0, 1971, 2441000, 74.1, Decimal('37966.00'), Decimal('36846.00'), 'Al-Imarat al-´Arabiya al-Muttahida', 'Emirate Federation', 'Zayid bin Sultan al-Nahayan', 65, 'AE'),
 ('ARG', 'Argentina', 'South America', 'South America', 2780400.0, 1816, 37032000, 75.1, Decimal('340238.00'), Decimal('323310.00'), 'Argentina', 'Federal Republic', 'Fernando de la Rúa', 69, 'AR'),
 ('ARM', 'Armenia', 'Asia', 'Middle East', 29800.0, 1991, 3520000, 66.4, Decimal('1813.00'), Decimal('1627.00'), 'Hajastan', 'Republic', 'Robert Kot\x9aarjan', 126, 'AM'),
 ('ABW', 'Aruba', 'North America', 'Caribbean', 193.0, None, 103000, 78.4, Decimal('828.00'), Decimal('793.00'), 'Aruba', 'Nonmetropolitan Territory of The Netherlands', 'Beatrix', 129, 'AW'),
 ('AUS', 'Australia', 'Oceania', 'Australia and New Zealand', 7741220.0, 1901, 18886000, 79.8, Decimal('351182.00'), Decimal('392911.00'), 'Australia', 'Constitutional Monarchy, Federation', 'Elisabeth II', 135, 'AU'),
 ('AZE', 'Azerbaijan', 'Asia', 'Middle East', 86600.0, 1991, 7734000, 62.9, Decimal('4127.00'), Decimal('4100.00'), 'Azärbaycan', 'Federal Republic', 'Heydär Äliyev', 144, 'AZ'),
 ('BHS', 'Bahamas', 'North America', 'Caribbean', 13878.0, 1973, 307000, 71.1, Decimal('3527.00'), Decimal('3347.00'), 'The Bahamas', 'Constitutional Monarchy', 'Elisabeth II', 148, 'BS'),
 ('BHR', 'Bahrain', 'Asia', 'Middle East', 694.0, 1971, 617000, 73.0, Decimal('6366.00'), Decimal('6097.00'), 'Al-Bahrayn', 'Monarchy (Emirate)', 'Hamad ibn Isa al-Khalifa', 149, 'BH'),
 ('BGD', 'Bangladesh', 'Asia', 'Southern and Central Asia', 143998.0, 1971, 129155000, 60.2, Decimal('32852.00'), Decimal('31966.00'), 'Bangladesh', 'Republic', 'Shahabuddin Ahmad', 150, 'BD'),
 ('BRB', 'Barbados', 'North America', 'Caribbean', 430.0, 1966, 270000, 73.0, Decimal('2223.00'), Decimal('2186.00'), 'Barbados', 'Constitutional Monarchy', 'Elisabeth II', 174, 'BB'),
 ('BEL', 'Belgium', 'Europe', 'Western Europe', 30518.0, 1830, 10239000, 77.8, Decimal('249704.00'), Decimal('243948.00'), 'België/Belgique', 'Constitutional Monarchy, Federation', 'Albert II', 179, 'BE'),
 ('BLZ', 'Beli

Use `*` with care, since this may cause an unnecessary burden to your network. 
Besides, it make your SQL code less readable. If someone is not highly familiar with the database, they might not know what is being returned. 


**Exercise 3: `LIMIT` keyword**

You can use the `LIMIT` keyword to limit the number of rows returned by the `SELECT` statement. 
Fill in the SQL query below to retrieve the `name` and `lifeexpectancy` columns from the `country` table, while limiting the number of rows to 10. 

In [8]:
%%sql 
/*
... name, ... 
   ... country
   LIMIT ...;
*/


SELECT name, lifeexpectancy
    FROM country
    LIMIT 10;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
10 rows affected.


[('Afghanistan', 45.9),
 ('Netherlands', 78.3),
 ('Netherlands Antilles', 74.7),
 ('Albania', 71.6),
 ('Algeria', 69.7),
 ('American Samoa', 75.1),
 ('Andorra', 83.5),
 ('Angola', 38.3),
 ('Anguilla', 76.1),
 ('Antigua and Barbuda', 70.5)]

**Exercise 4: Aliases**

As you can see in the previous exercise, `lifeexpectancy` is hard to read. 
We can rename the columns in the SELECT statement's output by using the `AS` keyword. 
This will not change the `country` table, only the output of the SELECT statement. 

Fill in the SQL query below to retrieve the `name` and `lifeexpectancy` columns from the country table. Rename `lifeexpectancy` to `life_expectancy`. Again, limit the output to 15 rows.

In [9]:
%%sql

/*
...
    name,
    ... ... life_expectancy
  FROM ...
  ...;
*/


SELECT
    name,
    lifeexpectancy AS life_expectancy
  FROM country
  LIMIT 15;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
15 rows affected.


[('Afghanistan', 45.9),
 ('Netherlands', 78.3),
 ('Netherlands Antilles', 74.7),
 ('Albania', 71.6),
 ('Algeria', 69.7),
 ('American Samoa', 75.1),
 ('Andorra', 83.5),
 ('Angola', 38.3),
 ('Anguilla', 76.1),
 ('Antigua and Barbuda', 70.5),
 ('United Arab Emirates', 74.1),
 ('Argentina', 75.1),
 ('Armenia', 66.4),
 ('Aruba', 78.4),
 ('Australia', 79.8)]

**Exercise 5: `DISTINCT`**

What are the continents in the present in the `country` table? We can add the `DISTINCT` modifier to the `SELECT` clause to return only distinct results.  

> Note: `DISTINCT` is applied to **all columns** that we list in front of `SELECT`, and returns all distinct combinations of values stored in those columns.

In [10]:
%%sql
/*
... DISTINCT ...
  FROM ...;
*/    

SELECT DISTINCT continent
  FROM country;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
7 rows affected.


[('Asia',),
 ('South America',),
 ('North America',),
 ('Oceania',),
 ('Antarctica',),
 ('Africa',),
 ('Europe',)]

**Exercise 6: Filtering with `WHERE`**

- The `WHERE` clause allows us to filter the rows we want;
- We can specify a condition (a Boolean expression) in the `WHERE` clause to retrieve only the rows for which the condition is true;
- Remember the filter verb in dplyr? It is the same idea!!

Retrieve the countries with a population higher than 150 million people. 


In [11]:
%%sql

/*
... ...
  FROM ...
  ... population > 150000000;
*/

SELECT name, population
  FROM country
  WHERE population > 150000000;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
6 rows affected.


[('Brazil', 170115000),
 ('Indonesia', 212107000),
 ('India', 1013662000),
 ('China', 1277558000),
 ('Pakistan', 156483000),
 ('United States', 278357000)]

- Some comparisons operator to use in `WHERE` clause

Operator | Condition |
---------|-----------|
`<`, `<=`, `=`, `>=`, `>` | ordinal comparisons | 
`LIKE`, `ILIKE`, `SIMILAR TO` | Pattern Matching | 
`BETWEEN` | Range filtering | 
`IN` | check belonginess (similar to `%in% `in R) |
`IS NULL` | check if is NULL |

**Exercise 7: derived columns**

SQL is able to make operations using columns. 

Retrieve the `name` column and `gnp_per_capita` by dividing the `gnp` column by the `population` column. 
Make sure to only calculate this for rows with population higher than 0. Remember to rename the column as `gnp_per_capita`. 

In [12]:
%%sql

/*
...
    name,
    ... ... gnp_per_capita
  FROM ...
  ... ...
  LIMIT 15;
*/


SELECT
    name,
    gnp / population AS gnp_per_capita
  FROM country
  WHERE population > 0
  LIMIT 15;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
15 rows affected.


[('Afghanistan', Decimal('0.00026302816901408451')),
 ('Netherlands', Decimal('0.02340910237014624307')),
 ('Netherlands Antilles', Decimal('0.00894470046082949309')),
 ('Albania', Decimal('0.00094231447724332588')),
 ('Algeria', Decimal('0.00158819230402592863')),
 ('American Samoa', Decimal('0.00491176470588235294')),
 ('Andorra', Decimal('0.02089743589743589744')),
 ('Angola', Decimal('0.00051622922814101569')),
 ('Anguilla', Decimal('0.00790000000000000000')),
 ('Antigua and Barbuda', Decimal('0.00900000000000000000')),
 ('United Arab Emirates', Decimal('0.01555346169602621876')),
 ('Argentina', Decimal('0.00918767552387124649')),
 ('Armenia', Decimal('0.00051505681818181818')),
 ('Aruba', Decimal('0.00803883495145631068')),
 ('Australia', Decimal('0.01859483215079953405'))]

> Remember:  the `SELECT` statement is powerful, but not dangerous. Derived columns returned by Postgres are not saved anywhere, nor do they change existing columns.

**Exercise 8: Arranging with `ORDER BY`**

What are the countries with the highest `gnp_per_capita`? 
You can use the `ORDER BY` clause to rearrange the rows that is returned. By default, it will arrange the rows in ascending order. 
If you want in descend order you can add the `DESC` keyword. 

Retrieve the `name` column and `gnp_per_capita` by dividing the `gnp` column by the `population` column. 
1. Make sure to only calculate this for rows with population higher than 0. 
2. Remember to rename the column as `gnp_per_capita`. 
3. Order the rows in descending order. 

In [13]:
%%sql

/*
...
    name,
    ... ... gnp_per_capita
  FROM ...
  ... population > 0
  ORDER BY ... DESC
  LIMIT 15;
*/


SELECT
    name,
    gnp / population AS gnp_per_capita
  FROM country
  WHERE population > 0
  ORDER BY gnp_per_capita DESC
  LIMIT 15;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
15 rows affected.


[('Luxembourg', Decimal('0.03745926095937571724')),
 ('Switzerland', Decimal('0.03693620468130272052')),
 ('Bermuda', Decimal('0.03581538461538461538')),
 ('Brunei', Decimal('0.03568597560975609756')),
 ('Liechtenstein', Decimal('0.03464396284829721362')),
 ('Cayman Islands', Decimal('0.03323684210526315789')),
 ('Denmark', Decimal('0.03266397748592870544')),
 ('Norway', Decimal('0.03257675561013732276')),
 ('United States', Decimal('0.03057476549898152373')),
 ('Japan', Decimal('0.02988653187493094686')),
 ('Iceland', Decimal('0.02958781362007168459')),
 ('Virgin Islands, British', Decimal('0.02914285714285714286')),
 ('Austria', Decimal('0.02618206085172643911')),
 ('Germany', Decimal('0.02596452004327892635')),
 ('Sweden', Decimal('0.02555939242106213465'))]

> Note: if there’s a draw in `ORDER BY`, the rows are presented in arbitrary order;


---  
**Remember:**

The order of SQL keywords does matter: `SELECT`, `FROM`, `WHERE`, `ORDER BY`, `LIMIT`
    
---

## Data Types

We will not cover data types in detail. You can find all data types supported by PostgreSQL [here](https://www.postgresql.org/docs/current/datatype.html).

One thing that is worth noting is the `CAST` function, to convert from one data type to another. 

**Example**

In [14]:
%%sql 

SELECT CAST('1.2587465416874' AS NUMERIC(5,2));

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
1 rows affected.


[(Decimal('1.26'),)]

### Nulls

- A null is marker to indicate that the value for a column is unknown, or not entered yet. A null is not equal to 0, or an empty string. In fact, **a null is not even equal to another null**!

- In `ORDER BY`, `NULL` values are considered either: the highest or the smallest possible value, depending on the DBMS. PostgreSQL treats `NULL` the highest possible value.

- Use `IS NULL` to match `NULL`. Do not try using equality. A `NULL` is not equal to another `NULL`.

- Although `NULL`s are not equal to each other, `DISTINC` treats them as if they were!

- How different environments show nulls:
  - `ipython-sql` -> `None`
  - psql -> blank space
  - pgAdmin -> `[null]`

------------------------

## `imdb` database 
Let's continue exploring the `WHERE` clause, but to keep things interesting, let's use another database. 

In [57]:
%config SqlMagic.displaylimit = 20

In [58]:
%sql postgresql://{username}:{password}@{host}:{port}/imdb

In [59]:
%%sql

/* Lets take a look at the movies table */

SELECT *
  FROM movies;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/imdb
   postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
26058 rows affected.


[(10035423, 'Kate & Leopold', None, 2001, None, 118, 6.4, 74982),
 (10042742, 'Mister 880', None, 1950, None, 90, 7.1, 1171),
 (10041181, 'Black Hand', None, 1950, None, 92, 6.4, 666),
 (10041387, 'Francis', None, 1950, None, 91, 6.4, 979),
 (10041719, 'Orpheus', 'Orphée', 1950, None, 95, 8.0, 9346),
 (10041931, 'Stromboli', 'Stromboli, terra di Dio', 1950, None, 107, 7.3, 5239),
 (10042052, 'Woman in Hiding', None, 1950, None, 92, 6.9, 553),
 (10042179, 'Abbott and Costello in the Foreign Legion', None, 1950, None, 80, 6.6, 2573),
 (10042200, 'Annie Get Your Gun', None, 1950, None, 107, 6.9, 4050),
 (10042206, 'Armored Car Robbery', None, 1950, None, 67, 7.0, 2077),
 (10042208, 'The Asphalt Jungle', None, 1950, None, 112, 7.9, 22106),
 (10042211, 'Atom Man vs. Superman', None, 1950, None, 252, 7.0, 579),
 (10042219, 'Backfire', None, 1950, None, 91, 6.6, 955),
 (10042229, 'The Baron of Arizona', None, 1950, None, 97, 7.0, 1748),
 (10042249, 'The Big Lift', None, 1950, None, 120, 6.5, 1266),
 (10042256, 'The Black Rose', None, 1950, None, 120, 6.3, 1657),
 (10042265, 'The Blue Lamp', None, 1950, None, 84, 6.9, 1426),
 (10042274, 'Borderline', None, 1950, None, 88, 6.1, 996),
 (10042275, 'Born to Be Bad', None, 1950, None, 94, 6.7, 1797),
 (10042279, 'Branded', None, 1950, None, 104, 6.7, 702),
 (10042285, 'Bright Leaf', None, 1950, None, 110, 6.6, 1033),
 (10042296, 'Caged', None, 1950, None, 96, 7.7, 3195),
 (10042315, 'The Cariboo Trail', None, 1950, None, 81, 6.0, 529),
 (10042325, 'Champagne for Caesar', None, 1950, None, 99, 7.4, 924),
 (10042332, 'Cinderella', None, 1950, None, 74, 7.3, 134235),
 (10042333, 'The Clouded Yellow', None, 1950, None, 95, 6.9, 609),
 (10042338, 'Colt .45', None, 1950, None, 74, 6.0, 758),
 (10042343, 'Convicted', None, 1950, None, 91, 6.9, 643),
 (10042344, 'Copper Canyon', None, 1950, None, 84, 6.3, 510),
 (10042352, 'Crisis', None, 1950, None, 95, 6.7, 1395),
 (10042355, 'Story of a Love Affair', 'Cronaca di un amore', 1950, None, 98, 7.1, 2209),
 (10042367, 'Cyrano de Bergerac', None, 1950, None, 113, 7.5, 3375),
 (10042372, 'Dallas', None, 1950, None, 94, 6.3, 1115),
 (10042376, "The Damned Don't Cry", None, 1950, None, 103, 7.2, 2114),
 (10042379, 'Dark City', None, 1950, None, 98, 6.8, 1650),
 (10042393, 'Destination Moon', None, 1950, None, 92, 6.3, 3388),
 (10042394, 'Destination Murder', None, 1950, None, 72, 6.0, 546),
 (10042395, "Devil's Doorway", None, 1950, None, 84, 7.3, 1200),
 (10042397, 'Dial 1119', None, 1950, None, 75, 6.7, 787),
 (10042447, 'Fancy Pants', None, 1950, None, 92, 6.6, 1050),
 (10042451, 'Father of the Bride', None, 1950, None, 92, 7.2, 9343),
 (10042464, 'The Flame and the Arrow', None, 1950, None, 88, 7.0, 3133),
 (10042488, 'The Fuller Brush Girl', None, 1950, None, 85, 6.7, 576),
 (10042490, 'The Furies', None, 1950, None, 109, 7.3, 2504),
 (10042509, 'The Glass Menagerie', None, 1950, None, 107, 7.1, 934),
 (10042513, 'Gone to Earth', None, 1950, None, 110, 7.1, 921),
 (10042524, 'The Great Rupert', None, 1950, None, 87, 6.6, 842),
 (10042530, 'Gun Crazy', None, 1950, None, 87, 7.6, 10362),
 (10042531, 'The Gunfighter', None, 1950, None, 85, 7.7, 8490),
 (10042542, 'The Happy Years', None, 1950, None, 110, 6.8, 609),
 (10042544, 'Harriet Craig', None, 1950, None, 94, 7.3, 1421),
 (10042546, 'Harvey', None, 1950, None, 104, 7.9, 50544),
 (10042579, 'House by the River', None, 1950, None, 83, 7.2, 2579),
 (10042593, 'In a Lonely Place', None, 1950, None, 94, 8.0, 22965),
 (10042609, 'The Jackie Robinson Story', None, 1950, None, 77, 6.4, 1004),
 (10042610, 'The Jackpot', None, 1950, None, 85, 6.9, 821),
 (10042619, 'Diary of a Country Priest', "Journal d'un curé de campagne", 1951, None, 115, 8.0, 8621),
 (10042634, 'Key to the City', None, 1950, None, 101, 6.4, 599),
 (10042644, 'Kim', None, 1950, None, 113, 6.6, 1630),
 (10042646, "King Solomon's Mines", None, 1950, None, 103, 6.8, 5301),
 (10042648, 'Kiss Tomorrow Goodbye', None, 1950, None, 102, 7.1, 1

### Pattern matching

It is a quite common situation that we want to find rows for which the values of one or more columns match a particular pattern. In SQL, this can be done either using `LIKE` or by using regular expressions. The syntax is as follows:

```sql
SELECT
    column1, column2
FROM
    table1
WHERE
    column1 [NOT] LIKE '<pattern>'
;
```

Postgres provides us with two wild-cards that we can use with `LIKE`:
- `%` matches any string of characters
- `_` matches a single character.

Pattern matching with `LIKE` is case sensitive; however, Postgres also provides the `ILIKE` keyword that has the same functionality as `LIKE` but is case-insensitive (not part of standard SQL).

> **Note:** With `LIKE` or `ILIKE`, the entire string should match the pattern.

---

**Example:** Retrieve those movies from the `movie` table whose title contains the word `'violin'` (note that `LIKE` is picky about letter cases in strings!)

---

In [62]:
%%sql

SELECT
    *
FROM
    movies
WHERE
    title ILIKE '%Violin%'
;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/imdb
   postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
5 rows affected.


[(10120802, 'The Red Violin', 'Le violon rouge', 1998, None, 130, 7.6, 30285),
 (10451966, 'The Violin', 'El violín', 2005, None, 98, 7.7, 2212),
 (12401715, "The Devil's Violinist", None, 2013, None, 122, 6.1, 3033),
 (14972904, 'The Violin Teacher', 'Tudo Que Aprendemos Juntos', 2015, None, 102, 6.8, 645),
 (10053987, 'The Steamroller and the Violin', 'Katok i skripka', 1961, None, 46, 7.5, 4867)]


**Exercise 9:** 

Retrieve those movies from the `movie` table whose title starts with the word `'Zero'`.

In [64]:
%%sql

SELECT
    *
FROM
    movies
WHERE
    title LIKE 'Zero%';

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/imdb
   postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
18 rows affected.


[(10095244, 'Zerograd', 'Gorod Zero', 1988, None, 103, 7.5, 1463),
 (10113557, 'Zero Kelvin', 'Kjærlighetens kjøtere', 1995, None, 118, 7.3, 1711),
 (10120906, 'Zero Effect', None, 1998, None, 116, 6.9, 13383),
 (10198837, 'Zero Tolerance', 'Noll tolerans', 1999, None, 108, 6.4, 3288),
 (10283693, 'Zero Woman: Red Handcuffs', 'Zeroka no onna: Akai wappa', 1974, None, 88, 6.6, 783),
 (10365960, 'Zero Day', None, 2002, None, 92, 7.2, 3840),
 (10421090, 'Zerophilia', None, 2005, None, 90, 6.2, 2177),
 (11592292, 'Zero 2', None, 2010, None, 90, 7.6, 5360),
 (11790885, 'Zero Dark Thirty', None, 2012, None, 157, 7.4, 254644),
 (12294965, 'Zero Charisma', None, 2013, None, 86, 6.2, 2384),
 (13576084, 'Zero Motivation', 'Efes beyahasei enosh', 2014, None, 97, 7.3, 3688),
 (15446858, 'Zero Days', None, 2016, None, 116, 7.8, 8081),
 (10051221, 'Zero Hour!', None, 1957, None, 81, 6.6, 1295),
 (10968712, 'Zero. Lilac Lithuania', 'Zero. Alyvine Lietuva', 2006, None, 82, 7.5, 2147),
 (11297858, 'Zero: An Investigation Into 9/11', None, 2008, None, 104, 7.8, 650),
 (11519661, 'Zero', None, 2009, None, 110, 6.8, 555),
 (12323372, 'Zero Point', 'Nullpunkt', 2014, None, 115, 7.4, 877),
 (16194850, 'Zero 3', None, 2017, None, 88, 6.6, 1094)]

**Exercise 10:** 

Retrieve those movies from the `movie` table whose title is 4 letters long and ends with the letter `'e'`.

In [65]:
%%sql

SELECT
    *
FROM
    movies
WHERE
    title LIKE '___e'
;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/imdb
   postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
71 rows affected.


[(10043539, 'Five', None, 1951, None, 93, 6.3, 1068),
 (10064694, 'More', None, 1969, None, 112, 6.5, 2155),
 (10066500, 'Hope', 'Umut', 1970, None, 100, 8.2, 2770),
 (10067814, 'Love', 'Szerelem', 1971, None, 88, 7.9, 1582),
 (10068306, 'Bone', None, 1972, None, 95, 6.8, 905),
 (10069158, 'Rage', None, 1972, None, 100, 6.3, 765),
 (10071803, 'Mame', None, 1974, None, 132, 6.1, 2490),
 (10080716, 'Fame', None, 1980, None, 134, 6.6, 18864),
 (10087182, 'Dune', None, 1984, None, 137, 6.5, 113255),
 (10088930, 'Clue', None, 1985, None, 94, 7.3, 71433),
 (10106438, 'Blue', None, 1993, None, 79, 7.3, 1778),
 (10112431, 'Babe', None, 1995, None, 91, 6.7, 110464),
 (10114323, 'Safe', None, 1995, None, 119, 7.2, 11475),
 (10116308, 'Fire', None, 1996, None, 108, 7.2, 5386),
 (10116722, 'Jude', None, 1996, None, 123, 6.9, 9547),
 (10123948, 'Cure', None, 1997, None, 111, 7.4, 8963),
 (10123755, 'Cube', None, 1997, None, 90, 7.2, 195849),
 (10123964, 'Life', None, 1999, None, 108, 6.8, 41588),
 (10188568, 'Rage', 'Do koske', 1997, None, 97, 6.9, 1623),
 (10217019, 'Sade', None, 2000, None, 100, 6.2, 1609),
 (10227005, 'Made', None, 2001, None, 95, 6.4, 16478),
 (10246072, 'Maze', None, 2000, None, 97, 6.3, 759),
 (10275719, 'Tape', None, 2001, None, 86, 7.3, 17850),
 (10308772, 'Pure', None, 2002, None, 96, 6.9, 2460),
 (10339079, 'Gone', None, 2002, None, 80, 7.1, 1048),
 (10347105, 'Fuse', 'Gori vatra', 2003, None, 105, 7.3, 1875),
 (10424448, 'Some', 'Sseom', 2004, None, 118, 6.1, 504),
 (10436724, 'Rize', None, 2005, None, 86, 7.1, 3050),
 (10484039, 'Glue', None, 2006, None, 110, 6.4, 1427),
 (10491399, 'Haze', None, 2005, None, 49, 6.6, 2179),
 (10497986, 'Time', 'Shi gan', 2006, None, 97, 7.2, 11216),
 (10807943, 'Dhee', None, 2007, None, 142, 7.4, 2350),
 (10814014, 'Apne', None, 2007, None, 173, 6.0, 2489),
 (10815138, 'Take', None, 2007, None, 98, 6.0, 1395),
 (10907657, 'Once', None, 2007, None, 86, 7.8, 105907),
 (11656190, 'Safe', None, 2012, None, 94, 6.5, 97531),
 (11996310, 'Lore', None, 2012, None, 109, 7.1, 13556),
 (12043979, 'Rose', 'Róza', 2011, None, 90, 7.6, 2861),
 (12093100, 'Home', 'Dom', 2011, None, 127, 7.5, 1065),
 (12202471, 'Love', 'Ai', 2012, None, 128, 6.6, 1348),
 (12224026, 'Home', None, 2015, None, 94, 6.6, 86931),
 (12948840, 'Life', None, 2015, None, 111, 6.0, 6756),
 (13153634, 'Hope', 'So-won', 2013, None, 122, 8.3, 4285),
 (13442006, 'Cake', None, 2014, None, 102, 6.4, 33466),
 (13716530, 'Elle', None, 2016, None, 130, 7.1, 55849),
 (13774694, 'Love', None, 2015, None, 135, 6.1, 38709),
 (13823098, 'Edie', None, 2017, None, 102, 6.5, 1071),
 (13850214, 'Dope', None, 2015, None, 103, 7.2, 77372),
 (14047846, 'Home', None, 2016, None, 103, 7.2, 838),
 (14384088, 'Rage', 'Ikari', 2016, None, 142, 7.1, 1881),
 (14741714, 'Five', None, 2016, None, 102, 6.3, 2252),
 (15638952, 'Dede', None, 2017, None, 97, 7.2, 788),
 (15927420, 'More', 'Daha', 2017, None, 115, 7.1, 1600),
 (16266538, 'Vice', None, 2018, None, 132, 7.2, 87744),
 (17207238, 'Jane', None, 2017, None, 90, 7.8, 3271),
 (17616148, 'Luce', None, 2019, None, 109, 7.1, 1196),
 (18638640, 'June', None, 2019, None, 141, 7.4, 829),
 (19412098, 'Fyre', None, 2019, None, 97, 7.2, 28255),
 (10106673, 'Dave', None, 1993, None, 110, 6.8, 46358),
 (10119092, 'Face', None, 1997, None, 105, 6.7, 4059),
 (11014762, 'Home', None, 2009, None, 120, 8.6, 20678),
 (11017456, 'Race', None, 2008, None, 149, 6.6, 11763),
 (11151923, 'Rage', 'Rabia', 2009, None, 89, 6.3, 1047),
 (11186357, 'Lope', None, 2010, None, 106, 6.0, 977),
 (11319569, 'Home', None, 2008, None, 98, 7.0, 3641),
 (11483753, 'Pure', 'Till det som är vackert', 2010, None, 97, 6.9, 3919),
 (12208144, 'Move', 'Drei Zimmer/Küche/Bad', 2012, None, 118, 6.7, 1236),
 (13499096, 'Race', None, 2016, None, 134, 7.1, 31582),
 (15442430, 'Life', None, 2017, None, 104, 6.6, 187432),
 (16342062, 'Anne', None, 2018, None, 80, 6.3, 606),
 (17715988, 'Cake', None, 2018, None, 125, 7.6, 753)]

## The relational model

SQL databases are based on the so-called relational model, it is based on the set theory in mathematics was introduced by by Edgar Codd (IBM) in 1970 ([more details here](https://en.wikipedia.org/wiki/Relational_model)). 
It's foundations in **set theory** is the reason you will here words like "tuples", "domain", "union", "cross product", etc.

We don't have time to cover the relational model in detail. But there are three things that you should keep in mind:


- **All operations** your perform in SQL result in a table (it could be an empty table).


- Primary keys: is one or more columns which uniquely identify each row (i.e., they are different for all rows); 
  - Examples: id, driver's license, SIN, license plate


- Foreign key: is one or more columns that uniquely identify each row of another table. For example
  - `id` in the `movies` table uniquely identify the movies;
  - `id` in the `names` table uniquely identify people; 
  - the `acting_roles` table relates people to movies using these primary keys to guarantee that there is no people working on a movie that doesn't exist, nor movies with people that don't exist (referential integrity);

In [73]:
%sql select * from names LIMIT 5;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/imdb
   postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
5 rows affected.


[(20000001, 'Fred Astaire', 1899, 1987),
 (20000002, 'Lauren Bacall', 1924, 2014),
 (20000003, 'Brigitte Bardot', 1934, None),
 (20000004, 'John Belushi', 1949, 1982),
 (20000005, 'Ingmar Bergman', 1918, 2007)]

In [75]:
%sql select * from movies LIMIT 5;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/imdb
   postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
5 rows affected.


[(10035423, 'Kate & Leopold', None, 2001, None, 118, 6.4, 74982),
 (10042742, 'Mister 880', None, 1950, None, 90, 7.1, 1171),
 (10041181, 'Black Hand', None, 1950, None, 92, 6.4, 666),
 (10041387, 'Francis', None, 1950, None, 91, 6.4, 979),
 (10041719, 'Orpheus', 'Orphée', 1950, None, 95, 8.0, 9346)]

In [74]:
%sql select * from acting_roles LIMIT 5;

 * postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/imdb
   postgresql://postgres:***@ssc-2022-workshop.ct6ghoz7smhy.us-east-1.rds.amazonaws.com:5432/world
5 rows affected.


[(10040559, 20685406, 'Celestino Esposito'),
 (10035423, 20000212, 'Kate McKay'),
 (10035423, 20413168, 'Leopold'),
 (10035423, 20000630, 'Stuart Besser'),
 (10035423, 20005227, 'Charlie McKay')]

There are also other useful models, which are less structured, which are known as `NoSQL` Databases. But this is outside the scope of this workshop.